modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/BaseUpgradeMatcherReplacementCheck.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.source.formatter.check;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Nícolas Moura
+ */
+public abstract class BaseUpgradeMatcherReplacementCheck extends BaseFileCheck {
+
+	protected String afterFormat(
+		String fileName, String absolutePath, String content,
+		String newContent) {
+
+		return newContent;
+	}
+
+	protected String beforeFormat(
+		String fileName, String absolutePath, String content) {
+
+		return content;
+	}
+
+	@Override
+	protected String doProcess(
+			String fileName, String absolutePath, String content)
+		throws Exception {
+
+		if (!isValidExtension(fileName)) {
+			return content;
+		}
+
+		String newContent = beforeFormat(fileName, absolutePath, content);
+
+		Matcher matcher = getPattern().matcher(content);
+
+		while (matcher.find()) {
+			newContent = format(content, newContent, matcher);
+		}
+
+		if (!content.equals(newContent)) {
+			newContent = afterFormat(
+				fileName, absolutePath, content, newContent);
+		}
+
+		return newContent;
+	}
+
+	protected abstract String format(
+		String content, String newContent, Matcher matcher);
+
+	protected abstract Pattern getPattern();
```

Comment:
Some of the upgrade checks needs to match multiple patterns. I think we need to change this to take multiple patterns instead.

We may need to create a new class that represents the pattern and the replacement. Something like `Replacement` or `UpgradeReplacement` (or something like that), where it will provide a `java.util.regex.Pattern` object and a `String` that will be used to do the replacement. It does not even have to be a `String`. For flexibility, we could turn it into something like a `Function<Matcher, String>` object.

So something like this:

```java
public interface Replacement {
    public Pattern getPattern();
    public Function<Matcher, String> getFunction();
}

...

public class BaseUpgradeMatcherReplacementCheck {

    @Override
    protected String doProcess(
		String fileName, String absolutePath, String content) {

        ...

        String newContent = content;

        for (Replacement replacement : getReplacements()) {
                Pattern pattern = replacement.getPattern();
                Matcher matcher = pattern.matcher(content);

                Function<Matcher, String> function = replacement.getFunction();
                newContent = StringUtil.replace(newContent, matcher.group(), function.apply(matcher));            
        }

        return newContent;
    }
    
    protected abstract Replacement[] getReplacements();
}
```

Something like this.

Reply:
We might need a new interface that is more complicated than `Function<Matcher, String>`.

Reply:
Just to make sure, are there any upgrade checks at the moment that have multiple patterns?

Reply:
@kevhlee We have 5 checks with multiple patterns but also some of them are more complex that will not be covered by a Base class like that. So the cases for a class like that are the minority right now. On the other hand, at least 8 checks already merged and 7 still to be merged are with a single pattern and will be perfectly served by this base class.